ksql-common/src/main/java/io/confluent/ksql/function/KsqlFunction.java

```diff
@@ -95,47 +136,30 @@ static KsqlFunction create(
         isVariadic);
   }
 
-  private KsqlFunction(
-      final Schema returnType,
-      final List<Schema> arguments,
-      final String functionName,
-      final Class<? extends Kudf> kudfClass,
-      final Function<KsqlConfig, Kudf> udfFactory,
-      final String description,
-      final String pathLoadedFrom,
-      final boolean isVariadic) {
-    this.returnType = Objects.requireNonNull(returnType, "returnType");
-    this.parameters = ImmutableList.copyOf(Objects.requireNonNull(arguments, "arguments"));
-    this.functionName = Objects.requireNonNull(functionName, "functionName");
-    this.kudfClass = Objects.requireNonNull(kudfClass, "kudfClass");
-    this.udfFactory = Objects.requireNonNull(udfFactory, "udfFactory");
-    this.description = Objects.requireNonNull(description, "description");
-    this.pathLoadedFrom  = Objects.requireNonNull(pathLoadedFrom, "pathLoadedFrom");
-    this.isVariadic = isVariadic;
-    this.hasGenerics = GenericsUtil.hasGenerics(returnType);
+  public Schema getReturnType(final List<Schema> arguments) {
 
-    if (arguments.stream().anyMatch(Objects::isNull)) {
-      throw new IllegalArgumentException("KSQL Function can't have null argument types");
+    final Schema returnType = returnSchemaProvider.apply(arguments);
+
+    if (returnType == null) {
+      throw new KsqlException("Return type of udf cannot be null.");
     }
-    if (isVariadic) {
-      if (arguments.isEmpty()) {
-        throw new IllegalArgumentException(
-            "KSQL variadic functions must have at least one parameter");
-      }
-      if (!Iterables.getLast(arguments).type().equals(Type.ARRAY)) {
-        throw new IllegalArgumentException(
-            "KSQL variadic functions must have ARRAY type as their last parameter");
+
+    //Cannot compare decimal types using equals as parameters won't match
+    if (DecimalUtil.isDecimal(returnType)) {
```

Comment:
```suggestion
    if (DecimalUtil.isDecimal(returnType) ^ DecimalUtil.isDecimal(javaReturnType)) {
```
Now we're getting fancy! In fact, this would generate a better error message where javaReturnType is decimal and returnType is not.

EDIT: actually, I think we should leverage the code in `UdfIndex.Parameter` class that compares java and schema types. That way we make sure that things also work for `Struct`. We should also add a unit test for the case where the schemaResolver returns some type of Struct.